src/codegen/compile.ml

```diff
@@ -10224,21 +10633,23 @@ and compile_prim_invocation (env : E.t) ae p es at =
   | NextArrayOffset spacing, [e] ->
     let advance_by =
       match spacing with
-      | ElementSize -> Arr.element_size
-      | One -> 2l (* 1 : Nat *) in
+      | ElementSize
+      | One -> Int32.shift_left 1l (32 - BitTagged.ubits_of Type.Int) (* 1 : Nat *) in
     SR.Vanilla,
     compile_exp_vanilla env ae e ^^ (* previous byte offset to array *)
     compile_add_const advance_by
   | ValidArrayOffset, [e1; e2] ->
     SR.bool,
-    compile_exp_vanilla env ae e1 ^^
-    compile_exp_vanilla env ae e2 ^^
+    compile_exp_vanilla env ae e1 ^^ BitTagged.untag_i32 __LINE__ env Type.Int ^^
+    compile_exp_vanilla env ae e2 ^^ BitTagged.untag_i32 __LINE__ env Type.Int ^^
     G.i (Compare (Wasm.Values.I32 I32Op.LtU))
   | DerefArrayOffset, [e1; e2] ->
     SR.Vanilla,
     compile_exp_vanilla env ae e1 ^^ (* skewed pointer to array *)
     Tagged.load_forwarding_pointer env ^^
     compile_exp_vanilla env ae e2 ^^ (* byte offset *)
```

Comment:
If I am not mistaken, the value would be the tagged index in the array:
```suggestion
    compile_exp_vanilla env ae e2 ^^ (* element index *)
```

Reply:
Fixed in #4375

Reply:
Fixed in #4375